File: sdks/typescript/src/serialize/responses.ts
/**
 * OpenAI Responses API codec — the ONE place the native model surface is
 * translated to and from the wire.
 *
 * Mirrors Rust `providers/responses.rs`. Both native providers
 * (providers/chatgpt_codex.ts and providers/openai.ts) call into here, so the
 * two cannot drift.
 *
 * Wire divergences that matter (all pinned by tests):
 *   - ModelToolCall.id            -> wire key `call_id` (decoding accepts `call_id` OR `id`)
 *   - ModelChatRequest.maxTokens  -> body key `max_output_tokens`
 *   - Tool.inputSchema            -> wire key `parameters`
 *   - freeform tools              -> wire `type: "custom"` (never "freeform")
 *   - freeform `input`            -> raw text, preserved byte-for-byte, NEVER JSON-parsed
 *
 * Unlike serialize/{anthropic,gemini,openai}.ts this module owns decoding and
 * the SSE adapter too, because the Responses codec is symmetric and shared.
 */

import type {
  FunctionCallOutputContentItem,
  FunctionCallOutputPayload,
  Message,
  ModelChatRequest,
  ModelContextItem,
  ModelToolCall,
  ModelToolOutput,
  ModelToolSpec,
  ToolChoice,
} from '../types.js'

/** Encode one tool definition into its Responses wire object. */
export function encodeToolSpec(spec: ModelToolSpec): Record<string, unknown> {
  if (spec.kind === 'freeform') {
    return {
      type: 'custom',
      name: spec.tool.name,
      description: spec.tool.description,
      format: {
        type: spec.tool.format.type,
        syntax: spec.tool.format.syntax,
        definition: spec.tool.format.definition,
      },
    }
  }
  // Rust's ToolSchema requires description/input_schema; TypeScript's Tool
  // makes them optional, so supply the empty forms rather than emitting null.
  return {
    type: 'function',
    name: spec.tool.name,
    description: spec.tool.description ?? '',
    parameters: spec.tool.inputSchema ?? {},
  }
}

/** Encode the whole tool list, order preserved. */
export function encodeTools(specs: ModelToolSpec[]): Record<string, unknown>[] {
  return specs.map(encodeToolSpec)
}

/** Encode a model tool call. `id` goes out under the wire key `call_id`. */
export function encodeToolCall(call: ModelToolCall): Record<string, unknown> {
  if (call.kind === 'freeform') {
    return {
      type: 'custom_tool_call',
      call_id: call.id,
      name: call.name,
      input: call.input,
    }
  }
  return {
    type: 'function_call',
    call_id: call.id,
    name: call.name,
    arguments: call.arguments,
  }
}

function encodeOutputContentItem(item: FunctionCallOutputContentItem): Record<string, unknown> {
  if (item.type === 'input_text') {
    return { type: 'input_text', text: item.text }
  }
  if (item.type === 'input_image') {
    const encoded: Record<string, unknown> = { type: 'input_image', image_url: item.imageUrl }
    if (item.detail !== undefined) encoded.detail = item.detail
    return encoded
  }
  return { type: 'encrypted_content', encrypted_content: item.encryptedContent }
}

function encodeOutputPayload(payload: FunctionCallOutputPayload): unknown {
  if (typeof payload === 'string') return payload
  return payload.map(encodeOutputContentItem)
}

/**
 * Encode a tool output. NOTE: the custom arm deliberately drops `name` — the
 * Responses body carries identity in `call_id` only, matching Rust
 * `encode_tool_output` (providers/responses.rs:34-49).
 */
export function encodeToolOutput(output: ModelToolOutput): Record<string, unknown> {
  if (output.kind === 'custom') {
    return {
      type: 'custom_tool_call_output',
      call_id: output.callId,
      output: encodeOutputPayload(output.output),
    }
  }
  return {
    type: 'function_call_output',
    call_id: output.callId,
    output: encodeOutputPayload(output.output),
  }
}

function encodeUserContent(message: Message): Record<string, unknown>[] {
  const blocks = message.contentBlocks ?? []
  if (blocks.length === 0) {
    return [{ type: 'input_text', text: message.content }]
  }

  const content: Record<string, unknown>[] = []
  for (const block of blocks) {
    if (block.type === 'text') {
      content.push({ type: 'input_text', text: block.text })
    } else if (block.type === 'image') {
      const source = block.source
      const imageUrl =
        source.type === 'base64'
          ? `data:${source.mediaType};base64,${source.data}`
          : source.url
      content.push({ type: 'input_image', image_url: imageUrl })
    }
    // Document blocks produce nothing (mirrors Rust ContentBlock::Document => {}).
  }

  if (content.length === 0) {
    content.push({ type: 'input_text', text: message.content })
  }
  return content
}

function encodeMessage(message: Message): Record<string, unknown>[] {
  switch (message.role) {
    case 'system':
      // System text belongs in `instructions`, never in `input`.
      return []
    case 'user':
      return [{ type: 'message', role: 'user', content: encodeUserContent(message) }]
    case 'assistant': {
      const items: Record<string, unknown>[] = []
      if (message.content) {
        items.push({
          type: 'message',
          role: 'assistant',
          content: [{ type: 'output_text', text: message.content }],
        })
      }
      for (const call of message.toolCalls ?? []) {
        items.push(
          encodeToolCall({
            kind: 'function',
            id: call.id,
            name: call.name,
            arguments: JSON.stringify(call.input) ?? '{}',
          }),
        )
      }
      return items
    }
    case 'tool':
      if (message.toolCallId === undefined) return []
      return [
        encodeToolOutput({
          kind: 'function',
          callId: message.toolCallId,
          output: message.content,
        }),
      ]
  }
}

/** Encode one ordered history entry into zero or more Responses input items. */
export function encodeContextItem(item: ModelContextItem): Record<string, unknown>[] {
  switch (item.kind) {
    case 'message':
      return encodeMessage(item.message)
    case 'toolCall':
      return [encodeToolCall(item.call)]
    case 'toolOutput':
      return [encodeToolOutput(item.output)]
  }
}

/** Encode the whole ordered context into the Responses `input` array. */
export function encodeInput(context: ModelContextItem[]): Record<string, unknown>[] {
  return context.flatMap(encodeContextItem)
}

function encodeToolChoice(choice: ToolChoice): unknown {
  switch (choice.type) {
    case 'auto':
      return 'auto'
    case 'required':
      return 'required'
    case 'none':
      return 'none'
    case 'tool':
      return { type: 'function', name: choice.name }
  }
}

/**
 * Build a Responses request body from a native model request.
 *
 * Two non-obvious rules, both load-bearing (Rust build_model_request_body,
 * providers/responses.rs:59-140):
 *
 *   1. System text is HOISTED into `instructions` and REMOVED from `input`.
 *      Precedence: systemBlocks (joined) > system > any Role::System message
 *      inside `context`; hoisted context messages are appended after whichever
 *      of the first two applied. `defaultInstructions` fills in only when
 *      nothing at all was supplied.
 *   2. `providerOptions` is shallow-merged into the body root LAST, so a
 *      caller can override anything this function produced.
 */
export function buildModelRequestBody(
  req: ModelChatRequest,
  defaultModel: string,
  stream: boolean,
  defaultInstructions?: string,
): Record<string, unknown> {
  const model = req.model ?? defaultModel
  const inputContext: ModelContextItem[] = []
  const instructionsParts: string[] = []

  if (req.systemBlocks !== undefined) {
    for (const block of req.systemBlocks) {
      const trimmed = block.text.trim()
      if (trimmed) instructionsParts.push(trimmed)
    }
  } else if (req.system !== undefined) {
    const trimmed = req.system.trim()
    if (trimmed) instructionsParts.push(trimmed)
  }

  for (const item of req.context) {
    if (item.kind === 'message' && item.message.role === 'system') {
      const trimmed = item.message.content.trim()
      if (trimmed) instructionsParts.push(trimmed)
      continue
    }
    inputContext.push(item)
  }

  const body: Record<string, unknown> = {
    model,
    input: encodeInput(inputContext),
  }

  if (stream) body.stream = true

  const toolSpecs = req.toolSpecs ?? []
  if (toolSpecs.length > 0) body.tools = encodeTools(toolSpecs)

  const instructions =
    instructionsParts.length > 0 ? instructionsParts.join('\n\n') : defaultInstructions
  if (instructions !== undefined) body.instructions = instructions

  if (req.temperature !== undefined) body.temperature = req.temperature
  // Wire key divergence: maxTokens -> max_output_tokens.
  if (req.maxTokens !== undefined) body.max_output_tokens = req.maxTokens
  if (req.toolChoice !== undefined) body.tool_choice = encodeToolChoice(req.toolChoice)
  if (req.stopSequences !== undefined && req.stopSequences.length > 0) {
    body.stop = req.stopSequences
  }

  // LAST — providerOptions wins over everything above.
  if (req.providerOptions !== undefined) {
    for (const [key, value] of Object.entries(req.providerOptions)) {
      body[key] = value
    }
  }

  return body
}

File: sdks/python/motosan_ai/providers/responses.py
"""Shared OpenAI Responses codec for the native model API.

Port of Rust ``sdks/rust/src/providers/responses.rs``. Pure encoding and
decoding — no HTTP, no provider state. Consumed by
``motosan_ai/providers/openai.py`` (behind the Responses opt-in) and
``motosan_ai/providers/chatgpt_codex.py`` (native by default).

Normative contract: ``specs/types.md`` § Native Model API.
"""

from __future__ import annotations

import json
from collections.abc import Sequence
from typing import Any

from motosan_ai.types import (
    FreeformTool,
    FunctionCallOutputContentItem,
    FunctionCallOutputInputImage,
    FunctionCallOutputInputText,
    FunctionCallOutputPayload,
    FunctionCallOutputText,
    ImageBlock,
    ImageSourceBase64,
    ImageSourceUrl,
    Message,
    ModelContextItem,
    ModelContextMessage,
    ModelContextToolCall,
    ModelToolCall,
    ModelToolCallFreeform,
    ModelToolCallFunction,
    ModelToolOutput,
    ModelToolOutputCustom,
    ModelToolOutputFunction,
    ModelToolSpec,
    ModelToolSpecFreeform,
    Role,
    TextBlock,
    ToolChoice,
)


def encode_freeform_tool(tool: FreeformTool) -> dict[str, Any]:
    """Wire shape for a Freeform tool spec.

    The ``type: "custom"`` key is injected here — ``FreeformTool`` never
    stores it.
    """
    return {
        "type": "custom",
        "name": tool.name,
        "description": tool.description,
        "format": {
            "type": tool.format.type,
            "syntax": tool.format.syntax,
            "definition": tool.format.definition,
        },
    }


def encode_tool_spec(spec: ModelToolSpec) -> dict[str, Any]:
    if isinstance(spec, ModelToolSpecFreeform):
        return encode_freeform_tool(spec.tool)
    return {
        "type": "function",
        "name": spec.tool.name,
        "description": spec.tool.description,
        # Wire key differs from the field name: input_schema -> parameters.
        "parameters": spec.tool.input_schema,
    }


def encode_tools(specs: Sequence[ModelToolSpec]) -> list[dict[str, Any]]:
    return [encode_tool_spec(spec) for spec in specs]


def encode_tool_call(call: ModelToolCall) -> dict[str, Any]:
    """Wire shape for a tool call. ``id`` is spelled ``call_id`` on the wire."""
    if isinstance(call, ModelToolCallFreeform):
        return {
            "type": "custom_tool_call",
            "call_id": call.id,
            "name": call.name,
            # Verbatim. Never parsed as JSON, never written to "arguments".
            "input": call.input,
        }
    return {
        "type": "function_call",
        "call_id": call.id,
        "name": call.name,
        "arguments": call.arguments,
    }


def encode_function_call_output_content_item(
    item: FunctionCallOutputContentItem,
) -> dict[str, Any]:
    if isinstance(item, FunctionCallOutputInputText):
        return {"type": "input_text", "text": item.text}
    if isinstance(item, FunctionCallOutputInputImage):
        encoded: dict[str, Any] = {"type": "input_image", "image_url": item.image_url}
        if item.detail is not None:
            encoded["detail"] = item.detail.value
        return encoded
    return {"type": "encrypted_content", "encrypted_content": item.encrypted_content}


def encode_function_call_output_payload(payload: FunctionCallOutputPayload) -> Any:
    """Text payloads encode to a bare string; content payloads to a list."""
    if isinstance(payload, FunctionCallOutputText):
        return payload.text
    return [encode_function_call_output_content_item(item) for item in payload.items]


def encode_tool_output(output: ModelToolOutput) -> dict[str, Any]:
    """Wire shape for a tool output item.

    TRAP: the custom arm deliberately DROPS ``name``. Rust's codec
    ``encode_tool_output`` emits only ``type`` / ``call_id`` / ``output``, and
    ``responses_codec_encodes_function_and_custom_outputs`` asserts the wire
    item has no ``name``. Use ``tool_output_to_dict`` when call identity must
    survive a round trip.
    """
    encoded_output = encode_function_call_output_payload(output.output)
    if isinstance(output, ModelToolOutputCustom):
        return {
            "type": "custom_tool_call_output",
            "call_id": output.call_id,
            "output": encoded_output,
        }
    return {
        "type": "function_call_output",
        "call_id": output.call_id,
        "output": encoded_output,
    }


def tool_output_to_dict(output: ModelToolOutput) -> dict[str, Any]:
    """Identity-preserving encoding: keeps ``name`` on custom outputs.

    Mirrors Rust's ``impl Serialize for ModelToolOutput`` rather than the
    codec's ``encode_tool_output``, and round-trips through
    ``decode_tool_output``. Not used to build request bodies.
    """
    encoded = encode_tool_output(output)
    if isinstance(output, ModelToolOutputCustom) and output.name is not None:
        encoded["name"] = output.name
    return encoded


def encode_user_content(message: Message) -> list[dict[str, Any]]:
    if not message.content_blocks:
        return [{"type": "input_text", "text": message.content}]

    content: list[dict[str, Any]] = []
    for block in message.content_blocks:
        if isinstance(block, TextBlock):
            content.append({"type": "input_text", "text": block.text})
        elif isinstance(block, ImageBlock):
            source = block.source
            if isinstance(source, ImageSourceBase64):
                content.append(
                    {
                        "type": "input_image",
                        "image_url": f"data:{source.media_type};base64,{source.data}",
                    }
                )
            elif isinstance(source, ImageSourceUrl):
                content.append({"type": "input_image", "image_url": source.url})
        # Document blocks have no Responses representation and are skipped.

    if not content:
        content.append({"type": "input_text", "text": message.content})
    return content


def encode_message(message: Message) -> list[dict[str, Any]]:
    """Expand one Message into zero or more Responses input items.

    System messages encode to nothing — ``build_model_request_body`` hoists
    them into ``instructions`` instead.
    """
    if message.role == Role.system:
        return []
    if message.role == Role.user:
        return [{"type": "message", "role": "user", "content": encode_user_content(message)}]
    if message.role == Role.assistant:
        items: list[dict[str, Any]] = []
        if message.content:
            items.append(
                {
                    "type": "message",
                    "role": "assistant",
                    "content": [{"type": "output_text", "text": message.content}],
                }
            )
        items.extend(
            encode_tool_call(
                ModelToolCallFunction(
                    id=tool_call.id,
                    name=tool_call.name,
                    arguments=json.dumps(tool_call.input),
                )
            )
            for tool_call in message.tool_calls
        )
        return items
    if message.tool_call_id is None:
        return []
    return [
        encode_tool_output(
            ModelToolOutputFunction(
                call_id=message.tool_call_id,
                output=FunctionCallOutputText(text=message.content),
            )
        )
    ]


def encode_context_item(item: ModelContextItem) -> list[dict[str, Any]]:
    if isinstance(item, ModelContextMessage):
        return encode_message(item.message)
    if isinstance(item, ModelContextToolCall):
        return [encode_tool_call(item.call)]
    return [encode_tool_output(item.output)]


def encode_input(context: Sequence[ModelContextItem]) -> list[dict[str, Any]]:
    encoded: list[dict[str, Any]] = []
    for item in context:
        encoded.extend(encode_context_item(item))
    return encoded


def encode_tool_choice(choice: ToolChoice) -> Any:
    if choice.type == "tool":
        return {"type": "function", "name": choice.name}
    return choice.type
